neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -246,57 +262,345 @@ func QmpQueryMemoryDevices(ip string, port int32) ([]QmpMemoryDevice, error) {
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleUp",
+			fmt.Sprintf("Added memslot%d to VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = false
+		r.maxBackend = max(r.maxBackend, idx)
+	}
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) AddDevices() bool {
+	attempts := NewAttemptsCounter(r.targetCnt - r.memDevCount)
+	for idx := 1; idx <= r.maxBackend; idx++ {
+		active, ok := r.memBackends[idx]
+		if !ok || active {
+			continue
+		}
+		// Found unused backend to plug into
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpAddMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleUp",
+				fmt.Sprintf("Failed to add dimm%d to VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleUp",
+			fmt.Sprintf("Added dimm%d to VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = true
+		r.memDevCount++
+	}
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveDevices() bool {
+	attempts := NewAttemptsCounter(r.memDevCount - r.targetCnt)
+	// Removing from the end to keep memslot1,memslot2,...
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || !active {
+			continue
+		}
+		// Found used backend to remove
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove dimm%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed dimm%d from VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = false
+		r.memDevCount--
+	}
+
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveBackends() bool {
+	attempts := NewAttemptsCounter(len(r.memBackends) - r.targetCnt)
+
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || active {
+			continue
+		}
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryBackend(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove memslot%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed memslot%d from VM %s", idx, r.vm.Name))
+		delete(r.memBackends, idx)
+	}
+	return attempts.DidSomething()
+}
+
+func QmpSetMemorySlots(ctx context.Context, vm *vmv1.VirtualMachine, targetCnt int, recorder record.EventRecorder) (int, error) {
+	// Attempts to plug/unplug memory slots to match targetCnt.
+	// Returns the number of slots, which the function managed to plug.
+	// Ideally, it matches targetCnt, but can be less or more if there are
+	// errors.
+	//
+	// Returns -1 if failed to connect to QMP, otherwise, the return value is
```

Comment:
This function is returning an error. The common Go practice to figure out if an error happened is to use the returned error value (if present). So I'd say in this case we can eliminate this guarantee (returning -1 if we failed) and stick to the returned error.

Reply:
I see, makes sense.